### PR TITLE
IMPRO-1331: Recursive filter modified to use symmetric padding

### PR DESCRIPTION
**Pull request by bayliffe**
This PR modifies the ```pad_cube_with_halo``` function to accept any np.pad method. It then uses this functionality in the ```RecursiveFilter``` plugin to implement the padding using the symmetric method.

- Unit tests are updated to reflect the change in values.
- Additional tests added to cover a few more pad options (non-exhaustive).
- KGO updated and plotted in associated notebook.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/utilities/test_pad_spatial.py  8
         

Clones removed
==============
+ improver_tests/nbhood/recursive_filter/test_RecursiveFilter.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5433319&bid=17817323)

File: improver_tests/nbhood/recursive_filter/test_RecursiveFilter.py [hunk 1]
```diff
@@ -352,50 +352,51 @@ def test_result_basic(self):
...
 
         expected_result = np.array(
             [
-                [0.01620921, 0.02866841, 0.05077430, 0.02881413, 0.01657352],
-                [0.03978802, 0.06457599, 0.10290188, 0.06486591, 0.04051282],
-                [0.10592333, 0.15184643, 0.19869247, 0.15238355, 0.10726611],
-                [0.03978982, 0.06457873, 0.10290585, 0.06486866, 0.04051464],
-                [0.01621686, 0.02868005, 0.05079120, 0.02882582, 0.01658128],
-            ]
+                [0.01320939, 0.02454378, 0.04346254, 0.02469828, 0.01359563],
```
Comment (fionaRust) [inline improver_tests/nbhood/recursive_filter/test_RecursiveFilter.py:392]: Without doing quite a bit of maths it is hard to say if these numbers are correct, but I don't see any reason why they shouldn't be.

File: improver/utilities/pad_spatial.py [hunk 2]
```diff
@@ -219,23 +219,17 @@ def pad_cube_with_halo(cube, width_x, width_y, halo_mean_data=True):
...
+    kwargs = {
+        "stat_length": ((width_y // 2, width_y // 2), (width_x // 2, width_x // 2))
+    }
+    if pad_method == "constant":
+        kwargs = {"constant_values": (0.0, 0.0)}
+    if pad_method == "symmetric":
+        kwargs = {}
+
+    padded_data = np.pad(
+        cube.data, ((width_y, width_y), (width_x, width_x)), pad_method, **kwargs
```
Comment (gavinevans) [inline improver/utilities/pad_spatial.py:231]: As `pad_method` is a keyword argument, it might be slightly better to do: `mode=pad_method`.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/utilities/test_pad_spatial.py  8
         

Clones removed
==============
+ improver_tests/nbhood/recursive_filter/test_RecursiveFilter.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5433319&bid=17817323)

File: improver/utilities/pad_spatial.py (hunk 2 above)
Comment (bayliffe) [reply improver/utilities/pad_spatial.py]: Agreed. Changed.

**Pull request merged**